src/plugins/mf_priority.cpp

```diff
@@ -0,0 +1,230 @@
+
+/* mycpppriority.cpp - custom basic job priority plugin
+ *
+ */
+
+extern "C" {
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+}
+#include <flux/core.h>
+#include <flux/jobtap.h>
+#include <map>
+#include <iterator>
+#include <cmath>
+#include <cassert>
+#include <algorithm>
+#include <cinttypes>
+
+std::map<int, std::map<std::string, double>> users;
+
+/******************************************************************************
+ *                                                                            *
+ *                           Helper Functions                                 *
+ *                                                                            *
+ *****************************************************************************/
+
+/*
+ * Calculate a user's job priority using the following factors:
+ *
+ * fairshare: the ratio between the amount of resources allocated vs. resources
+ *     consumed.
+ * urgency: a user-controlled factor to prioritize their own jobs.
+ */
+int64_t priority_calculation (flux_plugin_t *p,
+                              flux_plugin_arg_t *args,
+                              int userid,
+                              char *bank,
+                              int urgency)
+{
+    double fshare_factor = 0.0, priority = 0.0;
+    int fshare_weight;
+
+    std::map<int, std::map<std::string, double>>::iterator it;
+    std::map<std::string, double>::iterator inner_it;
+
+    fshare_weight = 100000;
+
+    if (urgency == FLUX_JOB_URGENCY_HOLD)
+        return FLUX_JOB_PRIORITY_MIN;
+
+    if (urgency == FLUX_JOB_URGENCY_EXPEDITE)
+        return FLUX_JOB_PRIORITY_MAX;
+
+    // search element in map of maps by key
+    it = users.find (userid);
+
+    // search for correct fshare value using passed-in bank; otherwise, use
+    // a default bank
+    if (bank != NULL) {
+        inner_it = it->second.find (bank);
+        fshare_factor = inner_it->second;
+    }
+    else {
+        inner_it = it->second.find ("default");
+        fshare_factor = inner_it->second;
+    }
+
+    priority = (fshare_weight * fshare_factor) + (urgency - 16);
+
+    return abs (round (priority));
+}
+
+
+/******************************************************************************
+ *                                                                            *
+ *                               Callbacks                                    *
+ *                                                                            *
+ *****************************************************************************/
+
+/*
+ * Unpack a payload from an external bulk update service and place it in the
+ * multimap datastructure.
+ */
+static void rec_update_cb (flux_t *h,
+                           flux_msg_handler_t *mh,
+                           const flux_msg_t *msg,
+                           void *arg)
+{
+    flux_plugin_t *p;
+    assert (sizeof arg == sizeof p);
+    memcpy (&p, &arg, sizeof arg);
```

Comment:
Thanks for letting me know about this. Just using:

```
flux_plugin_t *p = arg;
```

was giving me an error; I _think_ it was because the C++ compiler complains about an implicit conversion type to `void *`. I referenced [this](https://stackoverflow.com/questions/30088025/malloc-invalid-conversion-from-void-to-double) SO post. Let me know if I am way off base here. 

Regardless, I dropped those lines because, like you mentioned, they are not used anywhere in the function.

Reply:
Ah, sorry, C++ must be less forgiving of type conversion. You would want to do something like this instead:
```C
    flux_plugin_t *p = (flux_plugin_t *) arg;
```

Though I didn't test that with a c++ compiler so I could be wrong again 😉